osu.Game/Overlays/OnlineBeatmapSet/PreviewButton.cs

```diff
@@ -0,0 +1,157 @@
+                if (value == playing) return;
+                playing = value;
+
+                if (progress == null) return;
+
+                if (Playing)
+                {
+                    icon.Icon = FontAwesome.fa_stop;
+                    progress.FadeIn(100);
+
+                    loadPreview();
+                    preview.Start();
+                }
+                else
+                {
+                    icon.Icon = FontAwesome.fa_play;
+                    progress.FadeOut(100);
+                    preview.Stop();
+                }
+            }
+        }
+
+        public PreviewButton()
+        {
+            Height = 42;
+
+            Children = new Drawable[]
+            {
+                bg = new Box
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = Color4.Black.Opacity(0.25f),
+                },
+                new Container
+                {
+                    Anchor = Anchor.BottomLeft,
+                    Origin = Anchor.BottomLeft,
+                    RelativeSizeAxes = Axes.X,
+                    Height = 3,
+                    Child = progress = new Box
+                    {
+                        RelativeSizeAxes = Axes.Both,
+                        Width = 0f,
+                        Alpha = 0f,
+                    },
+                },
+                icon = new SpriteIcon
+                {
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    Icon = FontAwesome.fa_play,
+                    Size = new Vector2(18),
+                    Shadow = false,
+                },
+            };
+
+            Action = () => Playing = !Playing;
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(OsuColour colours, AudioManager audio)
+        {
+            this.audio = audio;
+            progress.Colour = colours.Yellow;
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            if (Playing)
+            {
+                progress.Width = (float)(preview.CurrentTime / preview.Length);
+                if (preview.HasCompleted) Playing = false;
+            }
+        }
+
+        protected override void Dispose(bool isDisposing)
+        {
+            Playing = false;
+            base.Dispose(isDisposing);
+        }
+
+        protected override bool OnHover(InputState state)
+        {
+            bg.FadeColour(Color4.Black.Opacity(0.5f), 100);
+            return base.OnHover(state);
+        }
+
+        protected override void OnHoverLost(InputState state)
+        {
+            bg.FadeColour(Color4.Black.Opacity(0.25f), 100);
+            base.OnHoverLost(state);
+        }
+
+        private void loadPreview()
+        {
+            if (preview == null || preview.HasCompleted && BeatmapSet != null)
+            {
+                preview = audio.Track.Get(BeatmapSet.OnlineInfo.Preview);

```

Comment:
this needs to be done asynchronously. currently it causes hard blocking on the update thread.

i'd possibly consider only loading it when the play button is hit.